server/src/main/java/org/apache/iotdb/db/query/reader/chunk/MemPageReader.java

```diff
@@ -54,92 +51,71 @@ public MemPageReader(
   public BatchData getAllSatisfiedPageData(boolean ascending) throws IOException {
     TSDataType dataType = chunkMetadata.getDataType();
     BatchData batchData = BatchDataFactory.createBatchData(dataType, ascending, false);
-    while (timeValuePairIterator.hasNextTimeValuePair()) {
-      TimeValuePair timeValuePair = timeValuePairIterator.nextTimeValuePair();
+    for (int i = 0; i < tsBlock.getPositionCount(); i++) {
       if (valueFilter == null
           || valueFilter.satisfy(
-              timeValuePair.getTimestamp(), timeValuePair.getValue().getValue())) {
-        batchData.putAnObject(timeValuePair.getTimestamp(), timeValuePair.getValue().getValue());
+              tsBlock.getTimeColumn().getLong(i), tsBlock.getColumn(0).getObject(i))) {
+        batchData.putAnObject(
+            tsBlock.getTimeColumn().getLong(i), tsBlock.getColumn(0).getObject(i));
       }
     }
```

Comment:
we have already know the datatype, should not use batchData.putAnObject() method. We should use putXXX, like putInt.

Reply:
Fixed